File: 2023/04.27/서지훈_백준_25516_거리가_k이하인_트리_노드에서_사과_수확하기.java
import java.io.*;
import java.util.ArrayList;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    static ArrayList<Integer>[] adjacent;
    static boolean[] visited;
    static int[] apples;
    static int k, result;
    public static void dfs(int node, int distance){
        if(k > distance){
            for(Integer next_node : adjacent[node]){
                if(!visited[next_node]){
                    visited[next_node] = true;
                    result += apples[next_node];
                    dfs(next_node, distance+1);
                }
            }
        }
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());

        adjacent = new ArrayList[n];
        for(int i = 0; i < n; ++i){
            adjacent[i] = new ArrayList<>();
        }

        for(int i = 0; i < n-1; ++i){
            st = new StringTokenizer(br.readLine());
            int p = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());

            adjacent[p].add(c);
            adjacent[c].add(p);
        }

        apples = new int[n];
        st = new StringTokenizer(br.readLine());
        for(int i = 0; i < n; ++i){
            apples[i] = Integer.parseInt(st.nextToken());
        }


        visited = new boolean[n];
        visited[0] = true;

        result = apples[0];
        dfs(0, 0);

        bw.write(Integer.toString(result));

        bw.flush();
        bw.close();
        br.close();
    }
}

File: 2023/04.27/서지훈_백준_25516_거리가_k이하인_트리_노드에서_사과_수확하기.py
import sys
sys.setrecursionlimit(1000000)


def dfs(node, distance):
    if k > distance:
        for next_node in adjacent[node]:
            if not visited[next_node]:
                visited[next_node] = True
                global result
                result += apples[next_node]
                dfs(next_node, distance+1)


n, k = map(int, input().split())
adjacent = [[] for _ in range(n)]
for _ in range(n-1):
    p, c = map(int, sys.stdin.readline().split())
    adjacent[p].append(c)
    adjacent[c].append(p)
apples = list(map(int, sys.stdin.readline().split()))


visited = [False for _ in range(n)]
visited[0] = True
result = apples[0]

dfs(0, 0)

print(result)